File: 2022/day04.hs
-- Advent of Code 2022, day 4
--  Venanzio Capretta

module Main where

import System.Environment
import Data.List
import Data.Char

import Control.Applicative
import qualified Data.Map as M

import FunParser
import AoCTools

main :: IO ()
main = do
  args <- getArgs
  puzzle (head args)

puzzle :: String -> IO ()
puzzle fileName = do
  input <- readFile fileName
  let xs = parseAll pInput input
  putStrLn ("Part 1: " ++ show (part1 xs))
  putStrLn ("Part 2: " ++ show (part2 xs))

-- Parsing the input

type Range = (Int,Int)

pRange :: Parser Range
pRange = do start <- integer
            symbol "-"
            end <- integer
            return (start,end)

pData :: Parser (Range,Range)
pData = do first <- pRange
           symbol ","
           second <- pRange
           return (first,second)

pInput :: Parser [(Range,Range)]
pInput = pLines pData

-- Part 1

contained :: (Range,Range) -> Bool
contained ((s1,e1),(s2,e2)) = s1 >= s2 && e1<=e2

eitherContained :: (Range,Range) -> Bool
eitherContained (r1,r2) = contained (r1,r2) || contained (r2,r1)

part1 :: [(Range,Range)] -> Int
part1 = length . filter eitherContained

-- Part 2

overlap :: (Range,Range) -> Bool
overlap ((s1,e1),(s2,e2)) = not (e1<s2 || e2<s1)

part2 :: [(Range,Range)] -> Int
part2 = length . filter overlap

File: 2022/python/day04.py
# Advent of Code 2022, day 4
#   Venanzio Capretta

print("Advent of Code 2022, Day 4")

f = open("../input04")
input = f.read().strip()
f.close()

# Parsing the input

def parsePair(s):
  vs = s.replace('-',',').split(',')
  return ((int(vs[0]),int(vs[1])),(int(vs[2]),int(vs[3])))

pairRange = list(map(parsePair,input.splitlines()))

# Part 1

def rangeContained(r0,r1):
  return(r0[0] >= r1[0] and r0[1] <= r1[1])

contained = 0
for (r0,r1) in pairRange:
  if rangeContained(r0,r1) or rangeContained(r1,r0):
    contained += 1

print("Part 1: ")
print(contained)

# Part 2

def overlap(r0,r1):
  return(r0[0]<=r1[1] and r1[0]<=r0[1])

overlapping = 0
for (r0,r1) in pairRange:
  if overlap(r0,r1):
    overlapping += 1

print("Part 2: ")
print(overlapping)

